### ISSUE TITLE
Performance bottleneck in wireMail() function

### ISSUE THREAD

**Issue by thetuningspoon** [processwire/processwire-issues#498]
### Short description of the issue

`wireMail()` calls take ~7-10 seconds on our server environment to complete, vs. ~0.01 seconds for a regular php `mail()` call

### Expected behavior

I expected a `wireMail()` call to be comparable to a standard `mail()` call, given that the native `WireMail` class is using `mail()` behind the scenes.

### Actual behavior

Sending a single email takes nearly 10 seconds!

### Optional: Suggestion for a possible fix

@zoaked and I used `Debug::timer()` to track down the issue, finally narrowing down to `WireMailTools::new()`. This method contains a loop through all modules in the system in order to determine if there are any WireMail modules extending the base WireMail class. Removing this call cut our processing time down to .03s, shaving off about 9 seconds!

Luckily this method is hookable, so our temporary fix was to add the following hook in our init.php file to bypass the search entirely and tell wireMail() which module to use:

```
$this->wire()->addHookBefore('WireMailTools::new', function($e) {
	$e->replace = true;
	$e->return = $e->modules('WireMailPHPMailer'); // Specify the name of your WireMail module
});

```

I'm not sure what would be the best long term fix for this. Perhaps we could use a configurable select box on the base WireMail module where the module you want to use can be saved to the database for fast retrieval. WireMail sub-modules could then automatically update this config option upon installation.

I think another way to avoid this issue is to use the class you want to use explicitly instead of using the global `wireMail()` function. So you would do `new WireMail()` or `new WireMailPHPMailer()`

### Steps to reproduce the issue

Add the following to your template file to test the processing time:

```
$timerkey = Debug::timer();
wireMail('toaddress@example.com', '', 'Test email', 'Message');
echo Debug::timer($timerkey);
```

### Setup/Environment

* ProcessWire version: 3.0.62
* PHP version: 7

**Comment by ryancramerdesign** [comment]
I can add a $config setting where you can specify the WireMail module to use, which should be a simple fix. Your hook solution also looks like a good one. But it seems like the real issue here isn't the time it takes to load WireMail, but rather why it's taking 7-10 seconds for that WireMailTools::new() call. That's an insane amount of time, and it sounds like you've found a major bottleneck, which might not have anything to do with WireMail. The only problem is, I can't seem to duplicate here. Timing my WireMailTools::new() calls on a current production site are coming out at 0.1 seconds. There must be some other bottleneck or reason why it's taking so long on your system. Do you have any ideas on how we might narrow in on this one? The only thing I can guess is that wireClassParents() call in WireMailTools::new(), but I'm not seeing the bottleneck behavior here so not sure how to narrow in. Is there anything unusual about the environment or installation? Thanks.

**Comment by thetuningspoon** [comment]
Interesting. Let me do some more testing and see if I can narrow it down further.

**Comment by thetuningspoon** [comment]
@ryancramerdesign It does appear to be related to `wireClassParents()`

Here is a readout of how long each iteration of the loop takes for each module:

```
ProcessRecentPages: 0.1182
 ProcessLogin: 0.1119
 ProcessPageList: 0.1139
 ProcessPageTrash: 0.1103
 ProcessPageClone: 0.1099
 ProcessPageEditLink: 0.1169
 ProcessTemplate: 0.1088
 ProcessPageType: 0.1094
 ProcessPermission: 0.1196
 ProcessPageSort: 0.1173
 ProcessPageEdit: 0.1065
 ProcessPageView: 0.0000
 ProcessPageSearch: 0.1137
 ProcessPageLister: 0.1397
 ProcessModule: 0.1066
 ProcessRole: 0.1022
 ProcessPageEditImageSelect: 0.1089
 ProcessField: 0.1148
 ProcessHome: 0.0911
 ProcessPageAdd: 0.0955
 ProcessLogger: 0.0970
 ProcessList: 0.1088
 ProcessUser: 0.1095
 FieldtypeDatetime: 0.0000
 FieldtypePageTitle: 0.0000
 FieldtypeFieldsetTabOpen: 0.1145
 FieldtypeFile: 0.0000
 InputfieldRepeater: 0.1142
 FieldtypeRepeater: 0.0000
 FieldtypeImage: 0.0000
 FieldtypeModule: 0.0000
 FieldtypeInteger: 0.0000
 FieldtypeFloat: 0.0000
 FieldtypeFieldsetClose: 0.1152
 FieldtypePassword: 0.0000
 FieldtypeURL: 0.0000
 FieldtypeEmail: 0.0000
 FieldtypeTextarea: 0.0000
 FieldtypeCheckbox: 0.0000
 FieldtypeFieldsetOpen: 0.0000
 FieldtypeOptions: 0.0000
 FieldtypePage: 0.0000
 FieldtypeText: 0.0000
 PageRender: 0.0000
 JqueryWireTabs: 0.1062
 JqueryMagnific: 0.1055
 JqueryUI: 0.1036
 JqueryCore: 0.1034
 JqueryTableSorter: 0.1038
 PagePermissions: 0.0001
 SystemUpdater: 0.0000
 MarkupPagerNav: 0.1060
 MarkupAdminDataTable: 0.1046
 MarkupPageArray: 0.0000
 MarkupHTMLPurifier: 0.1011
 InputfieldSelectMultiple: 0.1107
 InputfieldPageAutocomplete: 0.0993
 InputfieldFieldset: 0.0989
 InputfieldFile: 0.1128
 InputfieldImage: 0.1119
 InputfieldCKEditor: 0.1198
 InputfieldTextarea: 0.0000
 InputfieldPassword: 0.1203
 InputfieldName: 0.1267
 InputfieldDatetime: 0.1209
 InputfieldSubmit: 0.1035
 InputfieldForm: 0.1064
 InputfieldCheckbox: 0.1175
 InputfieldPage: 0.1110
 InputfieldIcon: 0.1017
 InputfieldSelect: 0.0000
 InputfieldInteger: 0.1026
 InputfieldButton: 0.1008
 InputfieldPageTitle: 0.1135
 InputfieldFloat: 0.1148
 InputfieldSelector: 0.1127
 InputfieldAsmSelect: 0.1131
 InputfieldCheckboxes: 0.1058
 InputfieldRadios: 0.1001
 InputfieldPageListSelectMultiple: 0.0963
 InputfieldPageListSelect: 0.1021
 InputfieldText: 0.0000
 InputfieldHidden: 0.1125
 InputfieldPageName: 0.1240
 InputfieldURL: 0.1032
 InputfieldEmail: 0.1147
 InputfieldMarkup: 0.1091
 SessionLoginThrottle: 0.0000
 ProcessSessionDB: 0.1036
 SessionHandlerDB: 0.0000
 TextformatterEntities: 0.1073
 TextformatterNewlineBR: 0.1216
 AdminThemeDefault: 0.1262
 InputfieldDecimal: 0.0036
 FieldtypeDecimal: 0.0000
 ImportPagesCSV: 0.0034
 ProcessExportProfile: 0.0049
 FieldtypeRepeaterMatrix: 0.1088
 InputfieldRepeaterMatrix: 0.1113
 FieldtypeMultiplier: 0.0071
 InputfieldMultiplier: 0.0072
 FieldtypeTextareas: 0.0035
 InputfieldTextareas: 0.0033
 InputfieldTable: 0.0042
 FieldtypeTable: 0.0000
 UiBlocks: 0.0000
 AdminPageFieldEditLinks: 0.0031
 ProcessProfile: 0.1089
 ProcessPageFieldSelectCreator: 0.0041
 WireMailPHPMailer: 0.0254

 Total loop time: 7.6529
```

If I remove the call to wireClassParents() and change the code so that it simply looks for the string "WireMail" in the module class name, the problem is fixed:
```

foreach($modules as $module) {
			//$parents = wireClassParents("$module");
			if(strpos($module->className, 'WireMail') !== false && $modules->isInstalled("$module")) {
				$mail = $modules->get("$module");
				break;
			}
		}
```

Here is what I get after running the modified code:


 ```
ProcessRecentPages: 0.0000
 ProcessLogin: 0.0000
 ProcessPageList: 0.0000
 ProcessPageTrash: 0.0000
 ProcessPageClone: 0.0000
 ProcessPageEditLink: 0.0000
 ProcessTemplate: 0.0000
 ProcessPageType: 0.0000
 ProcessPermission: 0.0000
 ProcessPageSort: 0.0000
 ProcessPageEdit: 0.0000
 ProcessPageView: 0.0000
 ProcessPageSearch: 0.0000
 ProcessPageLister: 0.0000
 ProcessModule: 0.0000
 ProcessRole: 0.0000
 ProcessPageEditImageSelect: 0.0000
 ProcessField: 0.0000
 ProcessHome: 0.0000
 ProcessPageAdd: 0.0000
 ProcessLogger: 0.0000
 ProcessList: 0.0000
 ProcessUser: 0.0000
 FieldtypeDatetime: 0.0000
 FieldtypePageTitle: 0.0000
 FieldtypeFieldsetTabOpen: 0.0000
 FieldtypeFile: 0.0001
 InputfieldRepeater: 0.0000
 FieldtypeRepeater: 0.0000
 FieldtypeImage: 0.0000
 FieldtypeModule: 0.0000
 FieldtypeInteger: 0.0000
 FieldtypeFloat: 0.0000
 FieldtypeFieldsetClose: 0.0000
 FieldtypePassword: 0.0000
 FieldtypeURL: 0.0000
 FieldtypeEmail: 0.0000
 FieldtypeTextarea: 0.0000
 FieldtypeCheckbox: 0.0000
 FieldtypeFieldsetOpen: 0.0000
 FieldtypeOptions: 0.0000
 FieldtypePage: 0.0000
 FieldtypeText: 0.0000
 PageRender: 0.0000
 JqueryWireTabs: 0.0000
 JqueryMagnific: 0.0000
 JqueryUI: 0.0000
 JqueryCore: 0.0000
 JqueryTableSorter: 0.0000
 PagePermissions: 0.0000
 SystemUpdater: 0.0000
 MarkupPagerNav: 0.0000
 MarkupAdminDataTable: 0.0000
 MarkupPageArray: 0.0000
 MarkupHTMLPurifier: 0.0000
 InputfieldSelectMultiple: 0.0000
 InputfieldPageAutocomplete: 0.0000
 InputfieldFieldset: 0.0000
 InputfieldFile: 0.0000
 InputfieldImage: 0.0000
 InputfieldCKEditor: 0.0000
 InputfieldTextarea: 0.0000
 InputfieldPassword: 0.0000
 InputfieldName: 0.0000
 InputfieldDatetime: 0.0000
 InputfieldSubmit: 0.0000
 InputfieldForm: 0.0000
 InputfieldCheckbox: 0.0000
 InputfieldPage: 0.0000
 InputfieldIcon: 0.0000
 InputfieldSelect: 0.0000
 InputfieldInteger: 0.0000
 InputfieldButton: 0.0000
 InputfieldPageTitle: 0.0000
 InputfieldFloat: 0.0000
 InputfieldSelector: 0.0000
 InputfieldAsmSelect: 0.0000
 InputfieldCheckboxes: 0.0000
 InputfieldRadios: 0.0000
 InputfieldPageListSelectMultiple: 0.0000
 InputfieldPageListSelect: 0.0000
 InputfieldText: 0.0000
 InputfieldHidden: 0.0000
 InputfieldPageName: 0.0000
 InputfieldURL: 0.0000
 InputfieldEmail: 0.0000
 InputfieldMarkup: 0.0000
 SessionLoginThrottle: 0.0000
 ProcessSessionDB: 0.0000
 SessionHandlerDB: 0.0000
 TextformatterEntities: 0.0000
 TextformatterNewlineBR: 0.0000
 AdminThemeDefault: 0.0000
 InputfieldDecimal: 0.0000
 FieldtypeDecimal: 0.0000
 ImportPagesCSV: 0.0000
 ProcessExportProfile: 0.0000
 FieldtypeRepeaterMatrix: 0.0000
 InputfieldRepeaterMatrix: 0.0000
 FieldtypeMultiplier: 0.0000
 InputfieldMultiplier: 0.0000
 FieldtypeTextareas: 0.0000
 InputfieldTextareas: 0.0000
 InputfieldTable: 0.0000
 FieldtypeTable: 0.0000
 UiBlocks: 0.0000
 AdminPageFieldEditLinks: 0.0000
 ProcessProfile: 0.0000
 ProcessPageFieldSelectCreator: 0.0000
 WireMailPHPMailer: 0.0328

 Total loop time: 0.0329

 Total: 0.0428
```

Much better. I don't know if you would want to use string comparison here as a solution to the problem, since it would require developers to properly name their modules. I will see if I can dive into wireClassParents() a bit.

**Comment by ryancramerdesign** [comment]
Thanks for trying that. It's interesting that it takes almost as long for a single module in your case, as it does for all modules in my case, even on my 5yr old laptop. Would be interesting to see if anyone else is seeing similar results. I'm wondering if we might be seeing a very specific environment or version triggered bottleneck. As for this case, I think a string match is probably fine, I'm not aware of any WireMail modules that don't start with the text "WireMail". But that won't solve other cases where we need to find modules having a certain parent class. I think it's worth tracking down the issue because it is probably slowing down a lot of stuff in your installation, and any other installations where it appears.

**Comment by thetuningspoon** [comment]
I think I've got it. We have a couple of extra `spl_autoload_register()` functions in our project which are part of a module I built (one is defined in the module's init() and the other in its ready(). The first checks a /site/templates/model/ directory for classes that extend the Page class. The second recursively searches through another directory in /site/templates/ to find classes which form the building blocks of our front end UI (I have suspected that this might not be the most efficient method). 

Apparently php is running through all of these other autoloaders before it finds what it's looking for. I'm not sure why this would be the case since I would expect PW's autoloader to be the first to run, which would quickly locate the module class it's looking for.

Anyway, removing our custom autoloaders sped the loop time up to ~.2 seconds, which is much better, but still not exactly ideal. I like your idea of adding a variable to the config file to manually set the WireMail module, and then falling back to the auto-detection (either by string or as it currently is) if the config variable is not set.

I'd also appreciate any thoughts you might have on how to better integrate our autoloaders into PW.

**Comment by ryancramerdesign** [comment]
Glad you found what it was. I've pushed an update that optimizes the WireMail module detection, by now focusing only on modules beginning with the WireMail prefix. Also added a $config option so that you can optimize it even further: 
~~~~~
$config->wireMail('module', 'WireMailPHPMailer'); 
~~~~~
Regarding your class loader, I don't know enough about your specific instance to know exactly what to suggest. But it sounds like that class loader might not have adequate caching? You could always try to use ProcessWire's class loader, via the $classLoader API var. `$classLoader->addNamespace('MyNamespace', '/path/to/my-namespace/');` But if you've already got things built, chances are you want to stick with your own class loader, so I would just check that there is adequate caching in there, so that it doesn't have to start over on every call.

**Comment by thetuningspoon** [comment]
Thanks Ryan!

I will look into caching as well as using PW's class loader, since I would prefer not to reinvent the wheel if PW already has what I need.

**Comment by thetuningspoon** [comment]
@ryancramerdesign ,
Just a follow-up on the autoloader issue. I was able to use `$classLoader->addNamespace('ProcessWire', '/path/to/model/folder');` in my init() to replace the first autoloader. Very cool!

The second autoloader is a bit more complex since I wanted it to support the ability to group classes into subfolders, without the need to rename the class or add a namespace to match its location in the directory tree. I am using `glob()` in a while loop to achieve this (it first checks all first level folders, then all second level, and so forth). I am not sure what the industry standard is for this, or whether it is just not recommended at all. 

Re: caching, I'm not sure what you mean. Doesn't php automatically cache any class found by spl_autoload_register? Or are you suggesting some form of database caching between page loads?

At any rate, I was able to significantly improve things by simply adding a check at the top of my spl_autoload_register which confirms that the class ends with 'UI' before it attempts to search for it. That way I can be sure the looping logic won't run unnecessarily.
